File: algorithm/intersection-of-two-arrays/intersection-of-two-arrays.cpp
class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        map<int, int> intersection;
        vector<int> result;
        
        for (auto n: nums1) {
            if (intersection.find(n) == intersection.end()) {
                intersection[n] = 1;
            }
        }
        
        for (auto n: nums2) {
            if (intersection.find(n) != intersection.end() && intersection[n] == 1) {
                intersection[n] = 2;
                result.push_back(n);
            }
        }
        
        return result;
    }
};

File: algorithm/intersection-of-two-arrays/intersection-of-two-arrays.go
package intersection_of_two_arrays

func intersection(nums1 []int, nums2 []int) []int {
	seen := make(map[int]int8)
	for _, v := range nums1 {
		seen[v] |= 1
	}

	for _, v := range nums2 {
		seen[v] |= 2
	}

	res := make([]int, 0)

	for v, cnt := range seen {
		if cnt == 3 {
			res = append(res, v)
		}
	}

	return res
}
